nttest_avx.cpp

```diff
@@ -0,0 +1,387 @@
+#include <cstdint>
+
+#include <getopt.h>
+#include "nthash_avx.hpp"
+
+#define PROGRAM "nttest_avx"
+
+static const char VERSION_MESSAGE[] =
+PROGRAM " Version 1.0.0 \n"
+"Written by Roman Snytsar.\n"
+"Copyright 2018 Microsoft Corp\n";
+
+static const char USAGE_MESSAGE[] =
+"Usage: " PROGRAM " [OPTION]... QUERY\n"
+"Report bugs to https://github.com/bcgsc/ntHash/issues\n";
+
+namespace opt {
+	unsigned kmerLen = 50;
+	uint64_t nz;
+	bool fastq = false;
+}
+
+using namespace std;
+
+static const char shortopts[] = "k:";
+
+enum { OPT_HELP = 1, OPT_VERSION };
+
+static const struct option longopts[] = {
+	{ "kmer",	required_argument, NULL, 'k' },
+	{ "help",	no_argument, NULL, OPT_HELP },
+	{ "version",	no_argument, NULL, OPT_VERSION },
+	{ NULL, 0, NULL, 0 }
+};
+
+static const string itm[] = { "ntbase","nthash", "ntavx2", "ntavx232", "ntavx512", "ntavx532" };
+
+void getFtype(const char *fName) {
+	std::ifstream in(fName);
+	std::string hLine;
+	bool good = static_cast<bool>(getline(in, hLine));
+	in.close();
+	if (!good) {
+		std::cerr << "Error in reading file: " << fName << "\n";
+		exit(EXIT_FAILURE);
+	}
+	if (hLine[0] == '>')
+		opt::fastq = false;
+	else if (hLine[0] == '@')
+		opt::fastq = true;
+	else {
+		std::cerr << "Error in file format: " << fName << "\n";
+		exit(EXIT_FAILURE);
+	}
+}
+
+bool getSeq(std::ifstream &uFile, std::string &line) {
+	bool good = false;
+	std::string hline;
+	line.clear();
+	if (opt::fastq) {
+		good = static_cast<bool>(getline(uFile, hline));
+		good = static_cast<bool>(getline(uFile, line));
+		good = static_cast<bool>(getline(uFile, hline));
+		good = static_cast<bool>(getline(uFile, hline));
+	}
+	else {
+		do {
+			good = static_cast<bool>(getline(uFile, hline));
+			if (hline[0] == '>' && !line.empty()) break;// !line.empty() for the first rec
+			if (hline[0] != '>')line += hline;
+		} while (good);
+		if (!good && !line.empty())
+			good = true;
+	}
+	return good;
+}
+
+void hashSeqb(const string & seq) {
+	for (size_t i = 0; i < seq.length() - opt::kmerLen + 1; i++) {
+		if (NTC64(seq.c_str() + i, opt::kmerLen)) opt::nz++;
+	}
+}
+
+void hashSeqr(const string & seq) {
+	uint64_t fhVal, rhVal, hVal;
+	hVal = NTC64(seq.c_str(), opt::kmerLen, fhVal, rhVal);
+	if (hVal)opt::nz++;
+	for (size_t i = 1; i < seq.length() - opt::kmerLen + 1; i++) {
+		hVal = NTC64(seq[i - 1], seq[i - 1 + opt::kmerLen], opt::kmerLen, fhVal, rhVal);
+		if (hVal)opt::nz++;
+	}
+}
+
+void hashSeqAvx2(const string & seq) {
+	const char* kmerSeq = seq.data();
+
+	__m256i _nz = _mm256_setzero_si256();
+	__m256i _zero = _mm256_setzero_si256();
+
+	__m256i _k = _mm256_kmod3133_epu64(opt::kmerLen);
+
+	__m256i _fhVal, _rhVal, _hVal;
+
+	_hVal = _mm256_NTC_epu64(kmerSeq, opt::kmerLen, _k, _fhVal, _rhVal);
+
+	__m256i _isZero = _mm256_cmpeq_epi64(
+		_hVal,
+		_zero);
+
+	_nz = _mm256_sub_epi64(
+		_nz,
+		_mm256_xor_si256(
+			_isZero,
+			_isZero));
+
+	kmerSeq += 3;
+
+	size_t sentinel = seq.length() - opt::kmerLen;
```

Comment:
Expression should read `seq.length() - opt::kmerLen + 1` otherwise last character isn't hashed and final hash doesn't match the scalar version.